packages/cicero-server/app.js

```diff
@@ -162,14 +164,122 @@ app.post('/draft/:template', async function (req, httpResponse, next) {
     }
 });
 
+/**
+ * Handle POST requests to /invoke/:template
+ * The body of the POST should contain the params, data and state.
+ * The clause is created using the template and the data.
+ * The call returns the output of requested clause.
+ *
+ * Template
+ * ----------
+ * The template parameter is the name of a directory under CICERO_DIR that contains
+ * the template to use.
+ *
+ * Request
+ * ----------
+ * The POST body contains six properties:
+ *  - sample or data
+ *  - parameters
+ *  - clause name
+ *  - state
+ *  - currentTime
+ *  - utcOffset
+ *
+ * Response
+ * ----------
+ * Output of the given clause from contract
+ *
+ */
+app.post('/invoke/:template', async function(req, httpResponse, next) {
+
+    try {
+        const options = req.body.options ? req.body.options : {};
+        const currentTime = req.body.currentTime ? req.body.currentTime : new Date().toISOString();
+        const utcOffset = req.body.utcOffset ? req.body.utcOffset : new Date().getTimezoneOffset();
+
+        const engine = new Engine();
+        const clause = await initTemplateInstance(req, options);
+        let clauseName;
+        let params;
+        let state;
+
+        if (req.body.clauseName) {
+            clauseName = req.body.clauseName.toString();
+        } else  {
+            throw new Error('Missing clause name in /invoke body');
+        }
+
+        if (req.body.params) {
+            params = req.body.params;
+        } else {
+            throw new Error('Missing params in /invoke body');
+        }
+
+        if (req.body.sample) {
+            clause.parse(req.body.sample.toString(), currentTime, utcOffset);
+        } else if (req.body.data) {
+            clause.setData(req.body.data);
+        } else {
+            throw new Error('Missing sample or data in /invoke body');
+        }
+
+        if(req.body.state) {
+            state = req.body.state;
+        } else {
+            const initResult = await engine.init(clause, currentTime, utcOffset);
+            state = initResult.state;
+        }
+
+        const result = await engine.invoke(clause, clauseName, params, state, currentTime, utcOffset);
+        httpResponse.status(200).send(result);
+    } catch(err) {
+        httpResponse.status(400).send({error: err.message});
+    }
+});
+
+/**
+ * Helper function to determine whether the template is archived or not
+ * @param {string} templateName Name of the template
+ * @returns {boolean} True if the given template is a .cta file
+ */
+function isTemplateArchive(templateName) {
+    try {
+        fs.lstatSync(`${process.env.CICERO_DIR}/${templateName}.cta`).isFile();
+        return true;
+    } catch(err) {
+        return false;
+    }
+}
+
+/**
+ * Helper function to load a template from disk
+ * @param {string} templateName Name of the template
+ * @param {object} options an optional set of options
+ * @returns {object} The template instance object.
+ */
+async function loadTemplate(templateName, options) {
```

Comment:
What's the motivation between separating `loadTemplate` from `initTemplateInstance`.

Having a single function that can dispatch between `Template.fromArchive`, `Template.fromUrl`, and `Template.fromDirectory` sounds useful?

Reply:
Some methods (`archive`, `get` and `compile`) do not require initializing a new clause, just returning the loaded template is enough. So, I split them and using the appropriate one for each method. There is no code duplication since we are already using `loadTemplate()` while initializing the clause. But there was a problem, `fromUrl()` needs to be in the `loadTemplate()` function instead of `initTemplateInstance()`. Fixed at https://github.com/accordproject/cicero/pull/746/commits/8e56d2f5ea58d0babac1030613892216700da411.